### PR TITLE
fix: WithContext had no effect

### LINKED ISSUE #1336: WithContext has no effect
The `WithContext` request option is broken because it does not actually update the request. I'm working on a fix.

### PR DESCRIPTION
**Pull request by armsnyder**
Fixes #1336

### CODE REVIEW

**Review by FedorLap2006 — COMMENTED** [review]
Hi, thanks for outlining the problem. Haven't noticed myself.
One thing to note though, I think the solution with replacing `Do(req)` to `Do(cfg.Request)` would look more consistent.

**Comment by FedorLap2006** [comment]
> @FedorLap2006 I still prefer my first version, but I changed it to reassign `req` in the caller function instead. I didn't want to simply replace `cfg.Client.Do(req)` with `cfg.Client.Do(cfg.Request)` since there are more places that `req` is used.

Ideally all occurences should be replaced with the `cfg.Request`. Besides it's only a couple (`req.Header` & `newRestError`)

**Comment by armsnyder** [comment]
@FedorLap2006 Hi, I followed your suggestion and updated the code. ☺️ 

My two cents are the first version with `*cfg.Request = *cfg.Request.WithContext(ctx)` is the cleaner and more consistent option.

More consistent: All the other `RequestOption`'s are mutating the original request. It seems that `WithContext` should also mutate it.

Cleaner: It's better to have only one request in-scope in the `RequestWithLockedBucket` function, instead of keeping a reference to the old `req` before applying options. It is a long function (111 lines!), so it may not be obvious for future code changes whether `req` or `cfg.Request` should be used.

You have final say, of course. 😅

**Review by FedorLap2006 — APPROVED** [review]
LGTM!

**Comment by FedorLap2006** [comment]
Regarding the original solution. After some consideration I have to admit the original solution could be the right way to go. Stylistically, it doesn't match up with other options, and can be a little bit confusing when reading `RequestWithLocketBucketID`, with all `cfg.XXX` around it. But logically, since other options mutate the original value (ref: `WithHeader`), it does make sense.

As for the reassignment of `req` inside `RequestWithLockedBucketID`, it doesn't match up stylistically and logicially with other options at all, that's why I've suggested to replace it by `cfg.Request`.

**Comment by FedorLap2006** [comment]
Sorry for taking matter into my own hands. After thinking for a while, the second solution (with reassigning `req`), does not affect readability that much. And has the least amount of changes.

P.S. Another way this could have been implemented was to use a `Context` field in `RequestConfig` and then reassign the `req` to `req.WithContext(cfg.Request)` (which is what I originally wanted to do), but `req = cfg.Request` would just be a simpler version of this.

File: restapi_test.go [hunk 1]
```diff
@@ -235,3 +237,41 @@ func Test_unmarshal(t *testing.T) {
...
+		return nil, testErr
+	})
+
+	// Run any client method using WithContext.
+	_, err = sess.User("", WithContext(ctx))
+
+	// Verify that the assertion code was actually run.
+	if !errors.Is(err, testErr) {
+		t.Errorf("unexpected error %v returned from client", err)
+	}
```
Comment (FedorLap2006) [inline restapi_test.go:269]:
```suggestion
	dg, err := New("")
	if err != nil {
		t.Fatal(err)
	}

	testErr := errors.New("test")

	// Intercept the request to assert the context.
	dg.Client.Transport = roundTripperFunc(func(r *http.Request) (*http.Response, error) {
		val, _ := r.Context().Value(key{}).(string)
		if val != "foo" {
			t.Errorf("missing value in context (got %q, wanted %q)", val, "foo")
		}
		return nil, testErr
	})

	// Run any client method using WithContext.
	_, err = dg.User("", WithContext(ctx))

	// Verify that the assertion code was actually run.
	if !errors.Is(err, testErr) {
		t.Errorf("unexpected error %v returned from client", err)
	}
```

File: restapi_test.go [hunk 2]
```diff
@@ -235,3 +237,41 @@ func Test_unmarshal(t *testing.T) {
 		t.Errorf("Unexpected error type: %T", err)
 	}
 }
+
+func TestWithContext(t *testing.T) {
+	// Set up a test context.
+	type key struct{}
+	ctx := context.WithValue(context.Background(), key{}, "foo")
```
Comment (FedorLap2006) [inline restapi_test.go:244]: I think it's going to be better to replace it with `value`, to be more consistent with the naming.

```suggestion
	ctx := context.WithValue(context.Background(), key{}, "value")
```

File: restapi_test.go [hunk 3]
```diff
@@ -235,3 +237,41 @@ func Test_unmarshal(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testErr := errors.New("test")
+
+	// Intercept the request to assert the context.
+	session.Client.Transport = roundTripperFunc(func(r *http.Request) (*http.Response, error) {
+		val, _ := r.Context().Value(key{}).(string)
+		if val != "foo" {
```
Comment (FedorLap2006) [inline restapi_test.go:257]: In relation to https://github.com/bwmarrin/discordgo/pull/1337/files#r1131680827
```suggestion
		if val != "value" {
```

**Comment by FedorLap2006** [comment]
Sorry for doing the commits right on the branch. Since the release is planned for today, I've decided to not prolong the things even more and apply all the edits myself.

Thank you for the contribution!

**Comment by armsnyder** [comment]
@FedorLap2006 No worries! Ultimately what matters most is the bug is fixed. 😄 Thanks for the review.

**Pull request merged**